clang-plugin/MozsearchIndexer.cpp

```diff
@@ -2591,6 +2597,46 @@ class IndexConsumer : public ASTConsumer,
     }
   }
 
+  bool arityMatchesCurrentCallExpr(const Expr *E, const NamedDecl *Candidate) {
+    const auto IsCurrentCallee = CurrentCall && E == CurrentCall->getCallee();
+    const auto CallNumArgs =
+        IsCurrentCallee ? CurrentCall->getNumArgs() : std::optional<uint>{};
+
+    const FunctionDecl *CandidateFunc;
+    if (const auto *UsingDecl = dyn_cast<UsingShadowDecl>(Candidate)) {
+      CandidateFunc = UsingDecl->getTargetDecl()->getAsFunction();
+    } else {
+      CandidateFunc = Candidate->getAsFunction();
+    }
+
+    // We try and filter candidates by arity, but be conservative and accept
+    // them when we don't know better
+    if (!CandidateFunc || !CallNumArgs) {
+      return true;
+    }
+
+    const auto MinNumArgs = CandidateFunc->getMinRequiredExplicitArguments();
```

Comment:
I've just checked this against:
```cpp
template <typename T>
struct Test {
  void method();
  void method(this Test<T> const &self);

  void test() {
    this->method();
  }
};

void test() {
  auto nonConstT = Test<int>{};
  nonConstT.method();

  const auto constT = Test<int>{};
  constT.method();
}

```

When dumping the values of all of those I get:
```
With unresolved callee in Test::test:
 CallExpr 0x1c23c408 '<dependent type>'
  `-UnresolvedMemberExpr 0x1c23c3a8 '<bound member function type>' lvalue
    `-CXXThisExpr 0x1c23c398 'Test<T> *' this
  getNumArgs: 0

  First candidate:
    CXXMethodDecl 0x1c23be58 <plugin-test-source/test.cpp:3:3, col:15> col:8 method 'void ()'
    getMinRequiredExplicitArguments: 0
    getMinRequiredArguments: 0
    getNumNonObjectParams: 0
    getNumParams: 0

  Second candidate:
    CXXMethodDecl 0x1c23c188 <plugin-test-source/test.cpp:4:3, col:39> col:8 method 'void (const Test<T> &)'
    `-ParmVarDecl 0x1c23c060 <col:15, col:35> col:35 self this 'const Test<T> &'
    getMinRequiredExplicitArguments: 0
    getMinRequiredArguments: 1
    getNumNonObjectParams: 0
    getNumParams: 1

With concrete callee in ::test:
  non-const version without deducing this:
    CXXMemberCallExpr 0x1c25efb0 'void'
    `-MemberExpr 0x1c25ef80 '<bound member function type>' .method 0x1c25e6f0
      `-DeclRefExpr 0x1c25ef00 'Test<int>':'struct Test<int>' lvalue Var 0x1c23c5d0 'nonConstT' 'Test<int>':'struct Test<int>'
    getNumArgs: 0

  const version with deducing this:
    CallExpr 0x1c2605e8 'void'
    |-ImplicitCastExpr 0x1c2605d0 'void (*)(const Test<int> &)' <FunctionToPointerDecay>
    | `-DeclRefExpr 0x1c260550 'void (const Test<int> &)' lvalue CXXMethod 0x1c25ea18 'method' 'void (const Test<int> &)'
    `-DeclRefExpr 0x1c25f428 'const Test<int>':'const struct Test<int>' lvalue Var 0x1c25f110 'constT' 'const Test<int>':'const struct Test<int>'
    getNumArgs: 1
```

So while `getNumArgs` would usually include the explicit object argument, _in our case_ it does not. It fact it can't because at that point we don't know yet which overload is going to be called.

Or am I missing other cases?

Reply:
Ah, good point! I overlooked the fact that the representation of a **dependent** call to a function with an explicit object argument is different, for exactly the reason you mention (we don't yet know which overload will be called). Thanks for checking.